Extract query loop product context validation into reusable hook

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

This PR adds a new hook `useQueryLoopProductContextValidation` that provides consistent validation for blocks that require a product context when used inside Query Loop blocks. The hook checks if the block is inside a Query Loop and validates that the post type is 'product', showing a warning message if the context is invalid.

Key changes:
- Adds new `useQueryLoopProductContextValidation` hook that can be used by any block requiring product context validation
- Updates Product Description block to use the new hook
- Updates Product Reviews block to use the new hook
- Exports the new hook from base hooks for reusability

This change improves code reusability and provides consistent validation behavior across blocks that require product context.

Closes #

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Prerequisites:**
1. Have the **WooCommerce Beta Tester** plugin activated
2. Go to **Tools > WCA Test Helper > Features** and enable the **experimental blocks** feature flag

**Steps:**
1. Create a new post 
2. Add Query Loop block
3. Add "Product Description" block inside "Post Template" block
4. Verify that you can see following warning message: "The Product Description block requires a product context. When used in a Query Loop, the Query Loop must be configured to display products."
![image](https://github.com/user-attachments/assets/c98100c7-857a-4e88-b539-7d5eef2daae5)

5. Set "Post type" to "Product" using Inspector control
![image](https://github.com/user-attachments/assets/6ab4c548-37a6-4831-9f1b-3927fc9a7494)

6. Verify that Product Description block is displayed correctly
7. Publish the post and view the frontend. Verify that Product Description block is displayed correctly.
8. Repeat steps 3-7 for "Blockified Product Reviews" block.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Extract query loop product context validation into reusable hook.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Initial commit for Product Description block.

This commit adds a new Product Description block to the WooCommerce Blocks plugin.
The block allows users to display and edit product descriptions in the block editor.

Most of the code was copied from Post Content block.

imanish003 committed:
Update product description block to handle product-specific content

imanish003 committed:
Remove console.log

imanish003 committed:
Implement product description block rendering on Frontend

This commit implements the rendering functionality for the product description block.
Key changes include:

- Add static tracking of seen product IDs to prevent recursive rendering
- Implement full render() method with proper product context handling
- Add safeguards against invalid products and empty descriptions
- Include proper block wrapper attributes and styling classes

Note: It's mostly inspired by Post Content block.

imanish003 committed:
Update block title

imanish003 committed:
Update product description block warning message

The warning message has been updated to be more descriptive and helpful,
explaining where the block can be used within product contexts.

imanish003 committed:
clean up TODOs in product description block

- Replace entry-content class with product-description__editable-content
- Rename isDescendentOfQueryLoop to isDescendentOfProductCollection
- Remove resolved TODOs and add clarifying comments

imanish003 committed:
Replace placeholder text with product description block

This commit replaces the placeholder text in the product details block with
the new product description block.

imanish003 committed:
Remove isDescendentOfProductCollection check from product description block

The isDescendentOfProductCollection check was preventing product descriptions
from being editable within product collection blocks. This was not the intended
behavior - product descriptions should be editable regardless of their location
in the block hierarchy.

imanish003 committed:
Move ProductDescription block within experimental blocks condition

imanish003 committed:
Add ancestor property and use registerProductBlockType

`registerProductBlockType` is used to register a block:
- On Single Product template, ancestor property is removed.
- Everywhere else, ancestor property is set to `['woocommerce/single-product', 'woocommerce/product-template']` for this block.

imanish003 committed:
Remove warning message from Product Description block

The warning message was removed as it's not needed anymore since the block
can only be inserted in product context.

imanish003 committed:
Merge branch 'trunk' into add/55723-product-details-block-create-product-description-block-3

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

imanish003 committed:
Add support for Product Description block in Query Loop

- Add core/post-template as valid ancestor in block.json
- Add validation to prevent usage in non-product Query Loops
- Add warning message when block is used in invalid context

imanish003 committed:
Implement HOC for Query Loop context validation

- Extract shared Query Loop context validation logic into a HOC
- Add Product Details block to product context only
- Update block registration to use registerProductBlockType
- Remove duplicate context validation code from Product Description block

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/product-details-available-in-product-context-only

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/product-details-available-in-product-context-only

imanish003 committed:
Minor refactor

imanish003 committed:
rename withInvalidQueryLoopContext to withQueryLoopProductContextValidation

This commit renames the HOC from withInvalidQueryLoopContext to withQueryLoopProductContextValidation
to better reflect its purpose. The component validates that product blocks are used in the correct
context - either outside a Query Loop or inside a Query Loop that queries products.

Additional changes:
- Improved TypeScript types for the HOC and its props
- Simplified the useSelect hook implementation
- Updated all imports to use the new name

imanish003 committed:
Remove changes related to product details block as it is being taken care in separate PR

imanish003 committed:
Move withQueryLoopProductContextValidation HOC to block-hocs

This commit moves the withQueryLoopProductContextValidation HOC from the
product-details block to the block-hocs directory, making it more accessible
for reuse across different blocks. The import paths have been updated
accordingly.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/product-details-available-in-product-context-only

imanish003 committed:
use withQueryLoopProductContextValidation HOC in Product Reviews block

The Product Reviews block now uses the withQueryLoopProductContextValidation HOC
to handle validation of the block's context within Query Loops. This replaces
the custom useIsInvalidQueryLoopContext hook with shared functionality.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/hocs/with-query-loop-product-context-validation.tsx:
```diff
@@ -0,0 +1,66 @@
+	WrappedComponent: React.ComponentType< EditComponentProps >,
+	blockName: string
+) => {
+	return function WithQueryLoopValidation( props: EditComponentProps ) {
```
Would an arrow function be more succinct here? Also do we need the function name?

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/hocs/with-query-loop-product-context-validation.tsx:
```diff
@@ -0,0 +1,66 @@
+) => {
+	return function WithQueryLoopValidation( props: EditComponentProps ) {
+		const { postType } = props.context;
+		const isInvalidQueryLoopContext = useSelect(
+			( select ) => {
+				const queryLoopAncestors = select(
+					blockEditorStore
+				).getBlockParentsByBlockName(
+					props.clientId,
+					'core/post-template'
+				);
+				return queryLoopAncestors.length > 0 && postType !== 'product';
+			},
+			[ props.clientId, postType ]
+		);
```
The old implementation used `useMemo`, just checking if this isn't needed anymore.

```
const { getBlockParentsByBlockName } = useSelect( blockEditorStore, [] );
	const blockParents = useMemo( () => {
		return getBlockParentsByBlockName( clientId, 'core/post-template' );
	}, [ getBlockParentsByBlockName, clientId ] );
```

louwie17 replied:
Correct, the old implementation also used `useSelect` in a different way where it just returned the selector name: `const { getBlockParentsByBlockName } = useSelect( blockEditorStore, [] );`
Instead of the recommended way of:
```
const value = useSelect( function( select ) {
  const selectors = select( store );
  const valueFromSelectors = selectors.selector(...);
  return valueFromSelectors;
}, [] );
```
The returned values within the `useSelect` function are memoized, so it uses `useMemo` internally.

cpapazoglou reviewed (dismissed):
Tested this and works as advertised:
Editor:
✅ Adding a product description in a posts query loops shows the error
![CleanShot 2025-03-19 at 14 25 11](https://github.com/user-attachments/assets/1167b451-644a-40d9-82bd-d235bc5ffc6c)
✅ Switching type to products displays the product description

Frontend:
✅ Works OK with a posts query loop
✅ Works OK with a products query loop
![CleanShot 2025-03-19 at 14 24 59](https://github.com/user-attachments/assets/c9aba5e4-d2fc-4365-8670-e38f7de8d91e)

Let's add a test for this HOC. I also added two minor comments.

louwie17 reviewed (dismissed):
I am not sure what our internal preference on this is, but I thought HOC's where not recommended anymore within React, especially since the hook era came along.
I did rather us just create a re-usable hook for this, or a re-usable component that we wrap within the Edit function. Instead of a specific HOC.

Also HOC's tend to turn into a TS nightmare, as it is easy for the context to get lost and in order to keep type safety, having to add generics in a lot of places.

imanish003 committed:
refactor: Convert product description block to use hook for context validation

This commit refactors the product description block to use a new hook
`useQueryLoopProductContextValidation` instead of the HOC
`withQueryLoopProductContextValidation`. This change:

- Extracts the query loop validation logic into a reusable hook
- Adds proper TypeScript types for the product description block
- Simplifies the component structure by removing the HOC wrapper
- Makes the validation logic more flexible and reusable across blocks

imanish003 committed:
Replace withQueryLoopProductContextValidation HOC with useQueryLoopProductContextValidation hook

This commit replaces the withQueryLoopProductContextValidation HOC with a hook-based approach using useQueryLoopProductContextValidation. This change:

- Removes the HOC implementation
- Updates the Product Reviews block to use the hook directly
- Maintains the same validation functionality for Query Loop context
- Simplifies the codebase by reducing HOC usage in favor of hooks

imanish003 commented:
> I am not sure what our internal preference on this is, but I thought HOC's where not recommended anymore within React, especially since the hook era came along. I did rather us just create a re-usable hook for this, or a re-usable component that we wrap within the Edit function. Instead of a specific HOC.
> 
> Also HOC's tend to turn into a TS nightmare, as it is easy for the context to get lost and in order to keep type safety, having to add generics in a lot of places.

Thank you for raising this concern. I agree with you. To be honest, I initially implemented a hook, but then we also needed a JSX for the warning message. I thought HOC would be better in this case. However, after reading your comment and thinking about it again, I have refactored the code and removed HOC. Instead, I am now using a hook for this. Let me know your thoughts on the changes. Do those look good to you?

cpapazoglou commented on plugins/woocommerce/changelog/56431-add-product-details-available-in-product-context-only:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: update
+
+Extract query loop product context validation into reusable HOC.
```
```suggestion
Extract query loop product context validation into a hook.
```

imanish003 replied:
@cpapazoglou Thanks for pointing this out. Actually I plan to update the PR description after receiving feedback from @louwie17, which will automatically update this change file. 🤝

louwie17 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-description/edit.tsx:
```diff
@@ -173,47 +178,22 @@ function RecursionError() {
+		useQueryLoopProductContextValidation( {
+			clientId,
+			postType: contextPostType,
+			blockName: 'Product Description',
```
Should this be translated?

imanish003 replied:
Agreed! Updated in [df30e31](https://github.com/woocommerce/woocommerce/pull/56431/commits/df30e313426936eb9561997494598a3321004bfc).

louwie17 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/edit.tsx:
```diff
@@ -50,22 +31,15 @@ const Edit = ( {
+		useQueryLoopProductContextValidation( {
+			clientId,
+			postType: context.postType,
+			blockName: 'Product Reviews',
```
Same as above, should this be translated?

imanish003 replied:
Agreed! Updated in [df30e31](https://github.com/woocommerce/woocommerce/pull/56431/commits/df30e313426936eb9561997494598a3321004bfc).

louwie17 reviewed (commented):
Thanks for the update @imanish003, I do like this hook approach better :)
Code changes look good, I just left two small comments around translating. 
I also didn't test it, but happy to do that as well if @cpapazoglou hasn't already?

cpapazoglou reviewed (approved):
Tested the same scenarios again and works well.

imanish003 committed:
Add translation support for block names

Commit:
Add changefile(s) from automation for the following project(s): woocommerce
